File: csv_editor.py
# csv_editor.py
import csv
import sys
import os
from typing import List, Dict, Any, Optional

class CSVEditor:
    def __init__(self, filename: Optional[str] = None):
        self.filename = filename
        self.headers: List[str] = []
        self.rows: List[List[str]] = []
        self.delimiter = ','
        if filename and os.path.exists(filename):
            self.load(filename)
        elif filename:
            # Create empty CSV with headers from user input
            self.create_empty()

    def load(self, filename: str):
        with open(filename, 'r', newline='', encoding='utf-8') as f:
            reader = csv.reader(f)
            rows = list(reader)
            if not rows:
                self.headers = []
                self.rows = []
                return
            self.headers = rows[0]
            self.rows = rows[1:]

    def save(self, filename: Optional[str] = None):
        if filename is None:
            filename = self.filename
        if not filename:
            filename = input("Enter filename to save: ").strip()
            if not filename.endswith('.csv'):
                filename += '.csv'
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(self.headers)
            writer.writerows(self.rows)

    def create_empty(self):
        print("Creating new CSV file.")
        headers = input("Enter column headers (comma-separated): ").strip()
        self.headers = [h.strip() for h in headers.split(',') if h.strip()]
        if not self.headers:
            self.headers = ['Column1', 'Column2', 'Column3']
        self.rows = []
        self.save()

    def display(self):
        if not self.headers:
            print("No data loaded.")
            return
        # Calculate column widths
        col_widths = [len(h) for h in self.headers]
        for row in self.rows:
            for i, cell in enumerate(row):
                if i < len(col_widths):
                    col_widths[i] = max(col_widths[i], len(str(cell)))
                elif i < len(self.headers):
                    col_widths.append(len(str(cell)))
        # Build header
        sep = ' | '
        header_line = sep.join(f"{h:<{col_widths[i]}}" for i, h in enumerate(self.headers))
        print("\n" + header_line)
        print('-' * len(header_line))
        for row in self.rows:
            cells = row + [''] * (len(self.headers) - len(row))
            print(sep.join(f"{cells[i]:<{col_widths[i]}}" for i in range(len(self.headers))))

    def add_row(self):
        if not self.headers:
            print("No headers defined. Please add columns first.")
            return
        row = []
        for i, h in enumerate(self.headers):
            val = input(f"Enter value for '{h}': ").strip()
            row.append(val)
        self.rows.append(row)
        print("Row added.")

    def edit_cell(self):
        self.display()
        try:
            row_idx = int(input("Enter row index (1-based): ")) - 1
            if row_idx < 0 or row_idx >= len(self.rows):
                print("Invalid row index.")
                return
            print("Columns:", ', '.join(self.headers))
            col_name = input("Enter column name: ").strip()
            if col_name not in self.headers:
                print("Column not found.")
                return
            col_idx = self.headers.index(col_name)
            new_val = input(f"Enter new value for '{col_name}': ").strip()
            self.rows[row_idx][col_idx] = new_val
            print("Cell updated.")
        except ValueError:
            print("Invalid input.")

    def delete_row(self):
        self.display()
        try:
            row_idx = int(input("Enter row index to delete (1-based): ")) - 1
            if row_idx < 0 or row_idx >= len(self.rows):
                print("Invalid row index.")
                return
            del self.rows[row_idx]
            print("Row deleted.")
        except ValueError:
            print("Invalid input.")

    def filter_rows(self):
        if not self.rows:
            print("No rows to filter.")
            return
        print("Columns:", ', '.join(self.headers))
        col_name = input("Enter column name to filter: ").strip()
        if col_name not in self.headers:
            print("Column not found.")
            return
        col_idx = self.headers.index(col_name)
        value = input("Enter value to match: ").strip()
        case_sensitive = input("Case-sensitive? (y/n): ").strip().lower() == 'y'
        filtered = []
        for row in self.rows:
            if col_idx < len(row):
                cell = row[col_idx]
                if case_sensitive:
                    match = cell == value
                else:
                    match = cell.lower() == value.lower()
                if match:
                    filtered.append(row)
        if not filtered:
            print("No matching rows.")
            return
        # Display filtered rows
        old_rows = self.rows
        self.rows = filtered
        self.display()
        self.rows = old_rows

    def sort_rows(self):
        if not self.rows:
            print("No rows to sort.")
            return
        print("Columns:", ', '.join(self.headers))
        col_name = input("Enter column name to sort by: ").strip()
        if col_name not in self.headers:
            print("Column not found.")
            return
        col_idx = self.headers.index(col_name)
        reverse = input("Sort descending? (y/n): ").strip().lower() == 'y'
        self.rows.sort(key=lambda row: row[col_idx] if col_idx < len(row) else '', reverse=reverse)
        print("Rows sorted.")

    def add_column(self):
        col_name = input("Enter new column name: ").strip()
        if not col_name:
            print("Invalid name.")
            return
        if col_name in self.headers:
            print("Column already exists.")
            return
        default_val = input("Enter default value for existing rows (leave empty for empty string): ").strip()
        self.headers.append(col_name)
        for row in self.rows:
            row.append(default_val)
        print("Column added.")

    def remove_column(self):
        if not self.headers:
            print("No columns to remove.")
            return
        print("Columns:", ', '.join(self.headers))
        col_name = input("Enter column name to remove: ").strip()
        if col_name not in self.headers:
            print("Column not found.")
            return
        col_idx = self.headers.index(col_name)
        del self.headers[col_idx]
        for row in self.rows:
            if col_idx < len(row):
                del row[col_idx]
        print("Column removed.")

    def summary(self):
        print(f"File: {self.filename or '(new)'}")
        print(f"Rows: {len(self.rows)}")
        print(f"Columns: {len(self.headers)}")
        if self.headers:
            print("Headers:", ', '.join(self.headers))

    def run(self):
        print("=== CSV Editor/Viewer ===")
        if self.filename:
            print(f"Loaded: {self.filename} ({len(self.rows)} rows, {len(self.headers)} columns)")
        else:
            print("No file loaded.")
        while True:
            print("\nMenu:")
            print("1. View table")
            print("2. Add row")
            print("3. Edit cell")
            print("4. Delete row")
            print("5. Filter rows")
            print("6. Sort rows")
            print("7. Add column")
            print("8. Remove column")
            print("9. Export / Save")
            print("10. Show summary")
            print("11. Exit")
            choice = input("Choose: ").strip()
            if choice == '1':
                self.display()
            elif choice == '2':
                self.add_row()
            elif choice == '3':
                self.edit_cell()
            elif choice == '4':
                self.delete_row()
            elif choice == '5':
                self.filter_rows()
            elif choice == '6':
                self.sort_rows()
            elif choice == '7':
                self.add_column()
            elif choice == '8':
                self.remove_column()
            elif choice == '9':
                self.save()
                print("Saved.")
            elif choice == '10':
                self.summary()
            elif choice == '11':
                if input("Save before exit? (y/n): ").strip().lower() == 'y':
                    self.save()
                print("Goodbye!")
                break
            else:
                print("Invalid choice.")

def main():
    filename = sys.argv[1] if len(sys.argv) > 1 else None
    editor = CSVEditor(filename)
    editor.run()

if __name__ == '__main__':
    main()

File: csv_editor.go
// csv_editor.go
package main

import (
	"bufio"
	"encoding/csv"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type CSVEditor struct {
	filename string
	headers  []string
	rows     [][]string
}

func NewCSVEditor(filename string) *CSVEditor {
	e := &CSVEditor{filename: filename}
	if filename != "" {
		if _, err := os.Stat(filename); err == nil {
			e.load(filename)
		} else {
			e.createEmpty()
		}
	} else {
		e.createEmpty()
	}
	return e
}

func (e *CSVEditor) load(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Println("Error loading file:", err)
		return
	}
	defer file.Close()
	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		fmt.Println("Error reading CSV:", err)
		return
	}
	if len(records) > 0 {
		e.headers = records[0]
		e.rows = records[1:]
	} else {
		e.headers = []string{}
		e.rows = [][]string{}
	}
}

func (e *CSVEditor) save(filename ...string) {
	fname := e.filename
	if len(filename) > 0 && filename[0] != "" {
		fname = filename[0]
	}
	if fname == "" {
		fmt.Print("Enter filename to save: ")
		scanner := bufio.NewScanner(os.Stdin)
		scanner.Scan()
		fname = strings.TrimSpace(scanner.Text())
		if !strings.HasSuffix(fname, ".csv") {
			fname += ".csv"
		}
	}
	file, err := os.Create(fname)
	if err != nil {
		fmt.Println("Error creating file:", err)
		return
	}
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()
	writer.Write(e.headers)
	for _, row := range e.rows {
		writer.Write(row)
	}
	fmt.Println("Saved.")
}

func (e *CSVEditor) createEmpty() {
	fmt.Println("Creating new CSV file.")
	fmt.Print("Enter column headers (comma-separated): ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	input := strings.TrimSpace(scanner.Text())
	if input == "" {
		e.headers = []string{"Column1", "Column2", "Column3"}
	} else {
		parts := strings.Split(input, ",")
		e.headers = make([]string, len(parts))
		for i, p := range parts {
			e.headers[i] = strings.TrimSpace(p)
		}
	}
	e.rows = [][]string{}
	e.save()
}

func (e *CSVEditor) display() {
	if len(e.headers) == 0 {
		fmt.Println("No data loaded.")
		return
	}
	// Calculate column widths
	colWidths := make([]int, len(e.headers))
	for i, h := range e.headers {
		colWidths[i] = len(h)
	}
	for _, row := range e.rows {
		for i, cell := range row {
			if i < len(colWidths) {
				if len(cell) > colWidths[i] {
					colWidths[i] = len(cell)
				}
			}
		}
	}
	// Build header
	sep := " | "
	fmt.Println()
	for i, h := range e.headers {
		fmt.Printf("%-*s", colWidths[i], h)
		if i < len(e.headers)-1 {
			fmt.Print(sep)
		}
	}
	fmt.Println()
	for i := range colWidths {
		fmt.Print(strings.Repeat("-", colWidths[i]))
		if i < len(colWidths)-1 {
			fmt.Print("-+-")
		}
	}
	fmt.Println()
	for _, row := range e.rows {
		cells := make([]string, len(e.headers))
		copy(cells, row)
		for i := range cells {
			fmt.Printf("%-*s", colWidths[i], cells[i])
			if i < len(cells)-1 {
				fmt.Print(sep)
			}
		}
		fmt.Println()
	}
}

func (e *CSVEditor) addRow() {
	if len(e.headers) == 0 {
		fmt.Println("No headers defined. Please add columns first.")
		return
	}
	row := make([]string, len(e.headers))
	for i, h := range e.headers {
		fmt.Printf("Enter value for '%s': ", h)
		scanner := bufio.NewScanner(os.Stdin)
		scanner.Scan()
		row[i] = strings.TrimSpace(scanner.Text())
	}
	e.rows = append(e.rows, row)
	fmt.Println("Row added.")
}

func (e *CSVEditor) editCell() {
	e.display()
	fmt.Print("Enter row index (1-based): ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	rowIdx, err := strconv.Atoi(strings.TrimSpace(scanner.Text()))
	if err != nil || rowIdx < 1 || rowIdx > len(e.rows) {
		fmt.Println("Invalid row index.")
		return
	}
	rowIdx--
	fmt.Println("Columns:", strings.Join(e.headers, ", "))
	fmt.Print("Enter column name: ")
	scanner.Scan()
	colName := strings.TrimSpace(scanner.Text())
	colIdx := -1
	for i, h := range e.headers {
		if h == colName {
			colIdx = i
			break
		}
	}
	if colIdx == -1 {
		fmt.Println("Column not found.")
		return
	}
	fmt.Printf("Enter new value for '%s': ", colName)
	scanner.Scan()
	newVal := strings.TrimSpace(scanner.Text())
	e.rows[rowIdx][colIdx] = newVal
	fmt.Println("Cell updated.")
}

func (e *CSVEditor) deleteRow() {
	e.display()
	fmt.Print("Enter row index to delete (1-based): ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	rowIdx, err := strconv.Atoi(strings.TrimSpace(scanner.Text()))
	if err != nil || rowIdx < 1 || rowIdx > len(e.rows) {
		fmt.Println("Invalid row index.")
		return
	}
	rowIdx--
	e.rows = append(e.rows[:rowIdx], e.rows[rowIdx+1:]...)
	fmt.Println("Row deleted.")
}

func (e *CSVEditor) filterRows() {
	if len(e.rows) == 0 {
		fmt.Println("No rows to filter.")
		return
	}
	fmt.Println("Columns:", strings.Join(e.headers, ", "))
	fmt.Print("Enter column name to filter: ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	colName := strings.TrimSpace(scanner.Text())
	colIdx := -1
	for i, h := range e.headers {
		if h == colName {
			colIdx = i
			break
		}
	}
	if colIdx == -1 {
		fmt.Println("Column not found.")
		return
	}
	fmt.Print("Enter value to match: ")
	scanner.Scan()
	value := strings.TrimSpace(scanner.Text())
	fmt.Print("Case-sensitive? (y/n): ")
	scanner.Scan()
	caseSensitive := strings.ToLower(strings.TrimSpace(scanner.Text())) == "y"
	var filtered [][]string
	for _, row := range e.rows {
		if colIdx < len(row) {
			cell := row[colIdx]
			match := false
			if caseSensitive {
				match = cell == value
			} else {
				match = strings.ToLower(cell) == strings.ToLower(value)
			}
			if match {
				filtered = append(filtered, row)
			}
		}
	}
	if len(filtered) == 0 {
		fmt.Println("No matching rows.")
		return
	}
	oldRows := e.rows
	e.rows = filtered
	e.display()
	e.rows = oldRows
}

func (e *CSVEditor) sortRows() {
	if len(e.rows) == 0 {
		fmt.Println("No rows to sort.")
		return
	}
	fmt.Println("Columns:", strings.Join(e.headers, ", "))
	fmt.Print("Enter column name to sort by: ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	colName := strings.TrimSpace(scanner.Text())
	colIdx := -1
	for i, h := range e.headers {
		if h == colName {
			colIdx = i
			break
		}
	}
	if colIdx == -1 {
		fmt.Println("Column not found.")
		return
	}
	fmt.Print("Sort descending? (y/n): ")
	scanner.Scan()
	reverse := strings.ToLower(strings.TrimSpace(scanner.Text())) == "y"
	sortFunc := func(i, j int) bool {
		a, b := "", ""
		if colIdx < len(e.rows[i]) {
			a = e.rows[i][colIdx]
		}
		if colIdx < len(e.rows[j]) {
			b = e.rows[j][colIdx]
		}
		if reverse {
			return a > b
		}
		return a < b
	}
	// Simple bubble sort for demonstration (avoid importing sort)
	for i := 0; i < len(e.rows)-1; i++ {
		for j := 0; j < len(e.rows)-i-1; j++ {
			if sortFunc(j, j+1) {
				e.rows[j], e.rows[j+1] = e.rows[j+1], e.rows[j]
			}
		}
	}
	fmt.Println("Rows sorted.")
}

func (e *CSVEditor) addColumn() {
	fmt.Print("Enter new column name: ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	colName := strings.TrimSpace(scanner.Text())
	if colName == "" {
		fmt.Println("Invalid name.")
		return
	}
	for _, h := range e.headers {
		if h == colName {
			fmt.Println("Column already exists.")
			return
		}
	}
	fmt.Print("Enter default value for existing rows (leave empty for empty string): ")
	scanner.Scan()
	defaultVal := strings.TrimSpace(scanner.Text())
	e.headers = append(e.headers, colName)
	for i := range e.rows {
		e.rows[i] = append(e.rows[i], defaultVal)
	}
	fmt.Println("Column added.")
}

func (e *CSVEditor) removeColumn() {
	if len(e.headers) == 0 {
		fmt.Println("No columns to remove.")
		return
	}
	fmt.Println("Columns:", strings.Join(e.headers, ", "))
	fmt.Print("Enter column name to remove: ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	colName := strings.TrimSpace(scanner.Text())
	colIdx := -1
	for i, h := range e.headers {
		if h == colName {
			colIdx = i
			break
		}
	}
	if colIdx == -1 {
		fmt.Println("Column not found.")
		return
	}
	e.headers = append(e.headers[:colIdx], e.headers[colIdx+1:]...)
	for i := range e.rows {
		if colIdx < len(e.rows[i]) {
			e.rows[i] = append(e.rows[i][:colIdx], e.rows[i][colIdx+1:]...)
		}
	}
	fmt.Println("Column removed.")
}

func (e *CSVEditor) summary() {
	fmt.Printf("File: %s\n", e.filename)
	fmt.Printf("Rows: %d\n", len(e.rows))
	fmt.Printf("Columns: %d\n", len(e.headers))
	if len(e.headers) > 0 {
		fmt.Println("Headers:", strings.Join(e.headers, ", "))
	}
}

func (e *CSVEditor) run() {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("=== CSV Editor/Viewer ===")
	if e.filename != "" {
		fmt.Printf("Loaded: %s (%d rows, %d columns)\n", e.filename, len(e.rows), len(e.headers))
	} else {
		fmt.Println("No file loaded.")
	}
	for {
		fmt.Println("\nMenu:")
		fmt.Println("1. View table")
		fmt.Println("2. Add row")
		fmt.Println("3. Edit cell")
		fmt.Println("4. Delete row")
		fmt.Println("5. Filter rows")
		fmt.Println("6. Sort rows")
		fmt.Println("7. Add column")
		fmt.Println("8. Remove column")
		fmt.Println("9. Export / Save")
		fmt.Println("10. Show summary")
		fmt.Println("11. Exit")
		fmt.Print("Choose: ")
		scanner.Scan()
		choice := strings.TrimSpace(scanner.Text())
		switch choice {
		case "1":
			e.display()
		case "2":
			e.addRow()
		case "3":
			e.editCell()
		case "4":
			e.deleteRow()
		case "5":
			e.filterRows()
		case "6":
			e.sortRows()
		case "7":
			e.addColumn()
		case "8":
			e.removeColumn()
		case "9":
			e.save()
		case "10":
			e.summary()
		case "11":
			fmt.Print("Save before exit? (y/n): ")
			scanner.Scan()
			if strings.ToLower(strings.TrimSpace(scanner.Text())) == "y" {
				e.save()
			}
			fmt.Println("Goodbye!")
			return
		default:
			fmt.Println("Invalid choice.")
		}
	}
}

func main() {
	var filename string
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}
	editor := NewCSVEditor(filename)
	editor.run()
}
